### ISSUE TITLE
Fix run_subset() accuracy placeholder in ValidationLoop

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#3680]
ValidationLoop.run_subset() still calls `metrics.update_val_metrics(avg_loss, 0.0)` with a hardcoded accuracy placeholder, mirroring the same bug fixed in run() by #3183. It should compute accuracy via AccuracyMetric over the subset batches (up to max_batches) and pass the real value. A follow-up test analogous to test_validation_loop_run_updates_val_accuracy() should verify metrics.val_accuracy > 0.0 after run_subset().

_Follow-up from #3183_

**Comment by mvillmow** [comment]
# Implementation Plan

Now I have a complete picture. Here's the implementation plan:

---

## Implementation Plan: Issue #3680 — Fix `run_subset()` accuracy placeholder

### 1. Objective

`ValidationLoop.run_subset()` passes a hardcoded `0.0` accuracy to `metrics.update_val_metrics()`. It must compute real accuracy using `AccuracyMetric` (mirroring the pattern in `run()`), but limited to `max_batches` batches. A new test `test_validation_loop_run_subset_updates_val_accuracy()` must verify `metrics.val_accuracy > 0.0` after calling `run_subset()`.

---

### 2. Approach

The fix mirrors exactly how `run()` computes accuracy (lines 216–223 of `validation_loop.mojo`), constrained to respect `max_batches`. Key decisions:

- Accumulate `AccuracyMetric` **inside** the existing loop (no second pass needed) — more efficient and simpler than a two-pass approach like `run()` uses.
- Guard the accuracy computation behind `if self.compute_accuracy:` to respect the flag (consistent with `run()`).
- The test uses the same helper functions (`simple_forward`, `create_val_loader`) and the same assertion pattern as `test_validation_loop_run_updates_val_accuracy()`.

---

### 3. Files to Create

None.

---

### 4. Files to Modify

#### `shared/training/loops/validation_loop.mojo`

**Location**: `run_subset()` method, lines 261–287.

**Change**: Add `AccuracyMetric` accumulation inside the batch loop, then pass the real accuracy to `update_val_metrics()`.

Current code:
```mojo
var total_loss = Float64(0.0)
var num_batches = 0

var loss_tracker = LossTracker(window_size=max_batches)

val_loader.reset()

while val_loader.has_next() and num_batches < max_batches:
    var batch = val_loader.next()

    var batch_loss = validation_step(
        model_forward, compute_loss, batch.data, batch.labels
    )

    loss_tracker.update(Float32(batch_loss))
    total_loss += batch_loss
    num_batches += 1

var avg_loss = Float64(0.0)
if num_batches > 0:
    avg_loss = total_loss / Float64(num_batches)
else:
    print("Warning: No validation batches available")

print("  Subset Validation Loss: " + String(avg_loss))

metrics.update_val_metrics(avg_loss, 0.0)  # ← BUG: hardcoded 0.0
```

New code:
```mojo
var total_loss = Float64(0.0)
var num_batches = 0

var loss_tracker = LossTracker(window_size=max_batches)
var accuracy_metric = AccuracyMetric()

val_loader.reset()

while val_loader.has_next() and num_batches < max_batches:
    var batch = val_loader.next()

    var batch_loss = validation_step(
        model_forward, compute_loss, batch.data, batch.labels
    )

    loss_tracker.update(Float32(batch_loss))
    total_loss += batch_loss
    num_batches += 1

    if self.compute_accuracy:
        var predictions = model_forward(batch.data)
        accuracy_metric.update(predictions, batch.labels)

var avg_loss = Float64(0.0)
if num_batches > 0:
    avg_loss = total_loss / Float64(num_batches)
else:
    print("Warning: No validation batches available")

print("  Subset Validation Loss: " + String(avg_loss))

var val_accuracy = Float64(0.0)
if self.compute_accuracy:
    val_accuracy = accuracy_metric.compute()

metrics.update_val_metrics(avg_loss, val_accuracy)

return avg_loss
```

Note: `model_forward` is called a second time per batch when computing accuracy. This mirrors the behavior of `run()` (which does a full second pass). The forward function is pure/stateless in tests.

#### `tests/shared/training/test_validation_loop.mojo`

**Location**: After `test_validation_loop_run_subset_resets_loader()` (around line 270), before the "No-Weight-Update Property Tests" section.

**Change**: Add one new test function:

```mojo
fn test_validation_loop_run_subset_updates_val_accuracy() raises:
    """Test run_subset() computes real accuracy via AccuracyMetric, not 0.0 placeholder.

    Follow-up to #3183 which fixed the same bug in run().
    Uses a 3-batch loader with max_batches=2 to verify the limit is respected
    and accuracy is non-zero after the call.
    """
    var vloop = ValidationLoop()
    var loader = create_val_loader(n_batches=3)
    var metrics = TrainingMetrics()
    _ = vloop.run_subset(simple_forward, simple_loss, loader, 2, metrics)
    assert_true(metrics.val_accuracy > 0.0)
    print("  test_validation_loop_run_subset_updates_val_accuracy: PASSED")
```

Also add a call to this function in the `main()` function at the bottom of the test file.

---

### 5. Implementation Order

1. **Read** `validation_loop.mojo` fully (already done — required before Write/Edit).
2. **Read** `test_validation_loop.mojo` fully (already done — required before Write/Edit).
3. **Edit** `shared/training/loops/validation_loop.mojo`: Replace the `run_subset()` body to accumulate `AccuracyMetric` and pass real accuracy to `update_val_metrics()`.
4. **Edit** `tests/shared/training/test_validation_loop.mojo`: Add `test_validation_loop_run_subset_updates_val_accuracy()` function body and call site in `main()`.
5. **Commit** with `SKIP=mojo-format` (GLIBC constraint on local host).
6. **Push** and create PR linked to `#3680`.
7. **Verify** in CI that the new test passes.

---

### 6. Verification

- **Test added**: `test_validation_loop_run_subset_updates_val_accuracy()` asserts `metrics.val_accuracy > 0.0`
- **Existing tests unchanged**: All 4 existing `run_subset` tests continue to pass (the change is additive within the method body)
- **Negative case**: The existing `test_validation_loop_run_compute_accuracy_false` pattern applies — if a `compute_accuracy=False` variant of the subset test is wanted, `val_accuracy` should remain `0.0`. This is covered implicitly by the `if self.compute_accuracy:` guard.
- **CI**: Push branch and confirm the `test-mojo` job passes (local GLIBC incompatibility prevents local execution).

---

### 7. Skills Used

**Planning phase (referenced prior learnings from team knowledge base)**:

| Skill | Source | Purpose |
|-------|--------|---------|
| `enable-disabled-mojo-tests` | Prior Learnings (team KB) | DataLoader 2D shape requirements, `create_val_loader` helper pattern, `SKIP=mojo-format` commit convention, Write-after-Read requirement |
| `validation-test-fixture-symmetry` | Prior Learnings (team KB) | Test symmetry between `run()` and `run_subset()` — new test mirrors `test_validation_loop_run_updates_val_accuracy()` |
| `evaluate-model` | Prior Learnings (team KB) | General accuracy/metrics evaluation patterns |

**Invoked directly during planning**:

- `Explore` subagent — read `ValidationLoop`, `AccuracyMetric`, and test file to establish exact line numbers and current state before drafting changes.

---
*Generated by Claude Code Planner*
